File: packages/typescript/examples/similarity_comparison.ts
#!/usr/bin/env ts-node
/**
 * Similarity comparison example.
 *
 * This example demonstrates:
 * - Comparing multiple vectors using LSH signatures
 * - Computing Hamming distance between signatures
 * - Estimating cosine similarity from Hamming distance
 *
 * Run with: npx ts-node typescript/examples/similarity_comparison.ts
 */

import {
  cosineFromHamming,
  hammingDistanceHex,
  normalizeVector,
  simhashLshMulti,
} from '../src/lsh';

function main() {
  console.log('=== LSH Similarity Comparison ===\n');

  // Three example vectors (unnormalized)
  const vectorA = [1.0, 1.0, 1.0, 1.0]; // Original
  const vectorB = [1.0, 0.9, 1.1, 1.0]; // Similar (small perturbation)
  const vectorC = [-1.0, -1.0, -1.0, -1.0]; // Opposite direction

  console.log('Input vectors:');
  console.log(`  A: ${JSON.stringify(vectorA)}`);
  console.log(`  B: ${JSON.stringify(vectorB)} (similar to A)`);
  console.log(`  C: ${JSON.stringify(vectorC)} (opposite to A)\n`);

  // Normalize all vectors
  const normA = normalizeVector(vectorA);
  const normB = normalizeVector(vectorB);
  const normC = normalizeVector(vectorC);

  // Generate signatures
  const sigA = simhashLshMulti(normA);
  const sigB = simhashLshMulti(normB);
  const sigC = simhashLshMulti(normC);

  console.log('Signatures (first family only):');
  console.log(`  A: ${sigA[0].signature.substring(0, 16)}`);
  console.log(`  B: ${sigB[0].signature.substring(0, 16)}`);
  console.log(`  C: ${sigC[0].signature.substring(0, 16)}`);
  console.log('     (showing first 16 hex chars of 64)\n');

  // Compare all pairs
  console.log('Pairwise comparisons:\n');

  compareSignatures('A vs B', sigA[0].signature, sigB[0].signature);
  compareSignatures('A vs C', sigA[0].signature, sigC[0].signature);
  compareSignatures('B vs C', sigB[0].signature, sigC[0].signature);

  console.log('\n✓ Comparison complete!');
  console.log('\nInterpretation:');
  console.log('  - Similarity > 0.9: Very similar');
  console.log('  - Similarity 0.7-0.9: Moderately similar');
  console.log('  - Similarity < 0.5: Dissimilar');
}

function compareSignatures(label: string, sig1: string, sig2: string): void {
  const hamming = hammingDistanceHex(sig1, sig2);
  const similarity = cosineFromHamming(hamming, 256);

  console.log(`${label}:`);
  console.log(`  Hamming distance: ${hamming}/256 bits differ`);
  console.log(`  Estimated cosine similarity: ${similarity.toFixed(4)}`);
  console.log();
}

main();

File: packages/rust/examples/similarity_comparison.rs
//! Similarity comparison example.
//!
//! This example demonstrates:
//! - Comparing multiple vectors using LSH signatures
//! - Computing Hamming distance between signatures
//! - Estimating cosine similarity from Hamming distance
//!
//! Run with: cargo run --example similarity_comparison

use odin_sig::{
    cosine_from_hamming, hamming_distance_hex, normalize_vector, simhash_lsh_multi, LshConfig,
};

fn main() {
    println!("=== LSH Similarity Comparison ===\n");

    // Three example vectors (unnormalized)
    let vector_a = vec![1.0, 1.0, 1.0, 1.0]; // Original
    let vector_b = vec![1.0, 0.9, 1.1, 1.0]; // Similar (small perturbation)
    let vector_c = vec![-1.0, -1.0, -1.0, -1.0]; // Opposite direction

    println!("Input vectors:");
    println!("  A: {:?}", vector_a);
    println!("  B: {:?} (similar to A)", vector_b);
    println!("  C: {:?} (opposite to A)\n", vector_c);

    // Normalize all vectors
    let norm_a = normalize_vector(&vector_a);
    let norm_b = normalize_vector(&vector_b);
    let norm_c = normalize_vector(&vector_c);

    // Generate signatures
    let config = LshConfig::default();
    let sig_a = simhash_lsh_multi(&norm_a, &config);
    let sig_b = simhash_lsh_multi(&norm_b, &config);
    let sig_c = simhash_lsh_multi(&norm_c, &config);

    println!("Signatures (first family only):");
    println!("  A: {}", &sig_a[0].signature[..16]);
    println!("  B: {}", &sig_b[0].signature[..16]);
    println!("  C: {}", &sig_c[0].signature[..16]);
    println!("     (showing first 16 hex chars of 64)\n");

    // Compare all pairs
    println!("Pairwise comparisons:\n");

    compare_signatures("A vs B", &sig_a[0].signature, &sig_b[0].signature);
    compare_signatures("A vs C", &sig_a[0].signature, &sig_c[0].signature);
    compare_signatures("B vs C", &sig_b[0].signature, &sig_c[0].signature);

    println!("\n✓ Comparison complete!");
    println!("\nInterpretation:");
    println!("  - Similarity > 0.9: Very similar");
    println!("  - Similarity 0.7-0.9: Moderately similar");
    println!("  - Similarity < 0.5: Dissimilar");
}

fn compare_signatures(label: &str, sig1: &str, sig2: &str) {
    let hamming = hamming_distance_hex(sig1, sig2);
    let similarity = cosine_from_hamming(hamming, 256);

    println!("{}:", label);
    println!("  Hamming distance: {}/256 bits differ", hamming);
    println!("  Estimated cosine similarity: {:.4}", similarity);
    println!();
}
